### PR TITLE
feature: req.headers.cookie / referer

### PR DESCRIPTION
**Pull request by Meemaw**
## What kind of change does this PR introduce?

Fixes https://github.com/toomuchdesign/next-page-tester/issues/25

## What is the current behaviour?

Cookie libraries check `document.cookie` when parsing cookies. We did not set that in our flow so authenticated pages would not work (depending on how you parsed cookies).

## What is the new behaviour?

`document.cookie` is correctly set which should correctly propagate to packages like `next-cookies`

## Does this PR introduce a breaking change?

No

## Please check if the PR fulfills these requirements:

- [x] Tests for the changes have been added
- [x] Docs have been added / updated

### CODE REVIEW

File: src/fetchData/makeHttpObjects.ts
Comment (toomuchdesign) [reply src/fetchData/makeHttpObjects.ts:21]: Since DOM tests should try to get as closer as possible to a realistic user interaction flow, It occurred to be that an user might want to:

- first, set desired cookies in JSDOM environment
- second, append those cookies in `req.headers.cookie`

Here it seems we're setting cookies in JSDOM for the user, but shouldn't a user prepare the JSDOM environment as expected before firing `getPage`?

Ironically we might also consider the opposite approach (even though it also has downsides): checking for `cookies` in `document` and append the to the `req` object. Probably this behaviour could have side-effects and could be optional.

Comment (toomuchdesign) [reply src/fetchData/makeHttpObjects.ts:21]: A third option could consist of leaving to the use decision/flexibility and add a few line of documentation about how to deal with `req` and cookies.

Comment (Meemaw) [reply src/fetchData/makeHttpObjects.ts:21]: My thoughts on this and library in general:

I'm thinking about this library on a bit higher level than JSDOM. We are rendering routes and not components, which is related to requests/responses in SSR NextJS.

TBH, I don't think I would consider setting `document.cookie` before rendering a route and expect that to work. Naturally I would expect this is done through the req headers.

But it's hard to argue what the best behaviour is without big usage. Current behaviour just seems most obvious to me ATM. Also, that is the last blocker for me to adopt this in a few larger projects and if everything goes smoothly start promoting this library a bit 😉

Comment (Meemaw) [reply src/fetchData/makeHttpObjects.ts:21]: Extra:

Perhaps current behaviour is okay only for initial page render, and should be adjusted for client navigations.

Comment (Meemaw) [reply src/fetchData/makeHttpObjects.ts:21]: Added an extra test to confirm client side login works (setting cookies on "Login" click)

File: src/fetchData/makeContextObject.ts
Comment (Meemaw) [reply src/fetchData/makeContextObject.ts:21]: Yep 🙂

File: src/fetchData/makeHttpObjects.ts
Comment (toomuchdesign) [reply src/fetchData/makeHttpObjects.ts:21]: `req.headers.cookie` mock is the cookie string, right?

**Review by toomuchdesign — APPROVED** [review]

File: src/fetchData/makeHttpObjects.ts
Comment (toomuchdesign) [reply src/fetchData/makeHttpObjects.ts:21]: > Perhaps current behaviour is okay only for initial page render, and should be adjusted for client navigations.

The issues I can currently see are:
- DOM cookies set after 1st rendering will be overridden e deleted by `req.headers.cookie`  at the following `getPage` call
- DOM cookies removed after 1st rendering will be recreated at the following `getPage` call

Did I get the context right?

File: src/__tests__/real-world-example/real-world-example.test.tsx
Comment (Meemaw) [inline src/__tests__/real-world-example/real-world-example.test.tsx:24]: Confirming cookies are not overwritten on page navigation. On page render we set only `'TrackingId=123';` and that is correctly propagated to the second page on client side navigation.

File: src/fetchData/fetchPageData.ts
Comment (Meemaw) [inline src/fetchData/fetchPageData.ts:123]: `getServerSideProps` always executes on server. For libraries are checking for `window` and `document` existence. For them to work as expected, we need to execute it in such context.

File: src/getPage.tsx
Comment (Meemaw) [inline src/getPage.tsx:86]: On client side navigation we are just passing forward the `document.cookie`. This is where we could also pass the `referer` header (this is how browser would behave).

File: src/loadPage.ts
Comment (Meemaw) [inline src/loadPage.ts:8]: No logic changes -- just a cleanup with reusing function

**Comment by Meemaw** [comment]
Hey @toomuchdesign . I did some more thinking and came with some new conclusions. New implementation should be more better and more accurate. I've added a few comments on the changes explaining.

**Review by toomuchdesign — COMMENTED** [review]
Hey @Meemaw,
I think we're gradually getting there. In order to better understand what we need to do, may I ask you to write down a brief list of the user scenarios we're trying to get covered?

PS. Shall we consider configuring cookies/`req.header.cookie` via a `cookie` option?

File: src/getPage.tsx
Comment (Meemaw) [reply src/getPage.tsx:86]: We could call user's `reqEnchancer` but override some of the values on our side when client side navigation happens:
- cookies
- referer
- ...

WDYT?

File: src/loadPage.ts
Comment (Meemaw) [reply src/loadPage.ts:7]: All seem very similar, but went with `3.` suggestion to hopefully make it a bit clearer.

File: src/__tests__/setup.ts
Comment (Meemaw) [reply src/__tests__/setup.ts:28]: Added a better comment.

Comment (toomuchdesign) [reply src/__tests__/setup.ts:29]: Silly question: what's the difference with setting `document.cookie` to `""`, which seems to be de default value?

File: src/getPage.tsx
Comment (toomuchdesign) [inline src/getPage.tsx:86]: Just one further consideration.
At the 1st render we assign  `document.cookie` the value found in `req.headers.cookie`.

From 2 render on, we're preserving cookies information by assigning `req.headers.cookie` `document.cookie` value ( in `makePage`).

So, if I get it right, we have `req.headers.cookie` -> `document.cookie`  -> `req.headers.cookie`.

I mentioned it a few commits ago, but it still seems a valid point to me. Would it make sense to make this flow unidirectional and make `document.cookie` the only source of true of cookie info?

This would involve mocking cookies in `document.cookie` instead of `req.headers.cookie`.

From my point of view it could make sense, since the cookies we're mocking into `req.headers.cookie` would be, in a real application, stored in `document.cookie` and later appended to a `fetch` call.

...or cookies can be handled differently in Next.js?

**Comment by Meemaw** [comment]
Okay, it seems I haven't used client side cookies enough to understand the API. So, it turns out to set the cookie you can set cookie like this:

```js
document.cookie = `aba=caba`;
```

You cannot set multiple cookies at once, but you need to call split them by semicolon and set them one by one.

Similarly, to delete the cookie, you need to do what we do in the `setup.ts` where we cleanup it. You need to set the cookie with an expiry date sometime in the past.

This means both of these don't work as you would expected:

```js
document.cookie = '';
document.cookie = someSuperValue; // this is not overriding cookie but is actually setting new cookie if needed
```

I have added some extra tests to confirm this behaviour and to verify it works correctly. Let me know if you see any case where the current implementation does not work.

**Comment by toomuchdesign** [comment]
Now that you mention it, I remember the awkward Cookie API. I've always used external libraries to avoid papercuts :)

File: src/__tests__/cookies/__fixtures__/with-ssr/pages/login.tsx
Comment (Meemaw) [inline src/__tests__/cookies/__fixtures__/with-ssr/pages/login.tsx:16]: Clear a cookie and make sure we are not setting it again on page navigation.

File: src/getPage.tsx
Comment (Meemaw) [inline src/getPage.tsx:86]: Call the user's reqEnahncer first. `req` name is confusing imo.

**Comment by Meemaw** [comment]
@toomuchdesign yeah exactly. That horrible API was usually hidden from me 🤯 

Anyway, I think I'm finally done on this PR. Sorry for a mess 💥  and thanks for bearing with me and making sure we implement this correctly 🙏

**Review by toomuchdesign — COMMENTED** [review]
@Meemaw thank you for bringing your take and your efforts.

1. Do you have in mind a better name for `req`/`res` options?
2. `isInitialRequest` being drilled down through different functions could be an indicator that there might be a different way to deal with "persisting" cookies.
3. What do you say about my previous comment?

> Just one further consideration.
> At the 1st render we assign `document.cookie` the value found in `req.headers.cookie`.
> 
> From 2nd render on, we're preserving cookies information by assigning `req.headers.cookie` `document.cookie` value ( in `makePage`).
> 
> So, if I get it right, we have `req.headers.cookie` -> `document.cookie` -> `req.headers.cookie`.
> 
> I mentioned it a few commits ago, but it still seems a valid point to me. Would it make sense to make this flow unidirectional and make `document.cookie` the only source of true of cookie info?
> 
> This would involve mocking cookies in `document.cookie` instead of `req.headers.cookie`.
> 
> From my point of view it could make sense, since the cookies we're mocking into `req.headers.cookie` would be, in a real application, stored in `document.cookie` and later appended to a `fetch` call.
> 
> ...or cookies can be handled differently in Next.js?

...providing a cookie string as `req` option is really better DX than providing a cookie object literal through a brand new `cookies` option or directly setting cookies in JSDOM before the `getPage` call?

🚀

**Comment by Meemaw** [comment]
> @Meemaw thank you for bringing your take and your efforts.
> 
> 1. Do you have in mind a better name for `req`/`res` options?

I think the external API is is okay with `req` and `res` (can't think of better names now; its short and concise), but we should map the variables to something more descriptive internally to avoid confusion (I've kept thinking they were `Request` and `Response` objects; thank god for Typescript though). I think `requestEnhancer`  and `responseEnhancer` would make more sense internally.

> 2. `isInitialRequest` being drilled down through different functions could be an indicator that there might be a different way to deal with "persisting" cookies.

The reason we currently have to do this is because `Request` object is created so deep in the call chain and there is no way to access passed cookies before that -- we only have a `Req => Req` function before that.

> 3. What do you say about my previous comment?
> 
> > Just one further consideration.
> > At the 1st render we assign `document.cookie` the value found in `req.headers.cookie`.
> > From 2nd render on, we're preserving cookies information by assigning `req.headers.cookie` `document.cookie` value ( in `makePage`).
> > So, if I get it right, we have `req.headers.cookie` -> `document.cookie` -> `req.headers.cookie`.
> > I mentioned it a few commits ago, but it still seems a valid point to me. Would it make sense to make this flow unidirectional and make `document.cookie` the only source of true of cookie info?
> > This would involve mocking cookies in `document.cookie` instead of `req.headers.cookie`.
> > From my point of view it could make sense, since the cookies we're mocking into `req.headers.cookie` would be, in a real application, stored in `document.cookie` and later appended to a `fetch` call.
> > ...or cookies can be handled differently in Next.js?
> 
> ...providing a cookie string as `req` option is really better DX than providing a cookie object literal through a brand new `cookies` option or directly setting cookies in JSDOM before the `getPage` call?
> 
> 🚀

If I understand your concern, you would like everything to be managed by `JSDOM` and `document.cookie` respectively. Unfortunately this is not possible in "server" environment. For example following wont work with SSR (or every data fetching method that should is executed on server by Next.JS).

Option 1 (won't work correctly with SSR):
```ts
// Setting cookies like this before a test also feels wierd
document.cookie = 'TrackingId=123';
document.cookie = 'AnalyticsId=12345';
document.cookie = 'Foo=Bar';

const { page } = await getPage({
  nextRoot: path.join(__dirname, '__fixtures__/with-ssr'),
  route: '/login',
});

// document.cookie is not available in `getServerSideProps` as `document` should not even be defined
// only way to read cookies in SSR is through req.headers
render(page); 
```

Options 2 (will work as expected in SSR):
```ts
const { page } = await getPage({
  nextRoot: path.join(__dirname, '__fixtures__/with-ssr'),
  route: '/login',
  req: (req) => {
    req.headers.cookie = 'TrackingId=123;AnalyticsId=12345;Foo=Bar';
    return req;
    },
});

render(page); 
```

After the initial render cookie handling is in full control of `JSDOM`. The only think we are doing is setting `req.headers.cookie = document.cookie` (browser would do that by default) so that cookies are available in following `getServerSideProps` executions caused by client side navigation.

**Comment by toomuchdesign** [comment]
Hi @Meemaw, sorry if I'm stressing much about this detail but, if i'm not overlooking anything, I believe is worth clarifying it.

I'm focusing on Option 1 (which I find very close to DOM testing philosophy).

I just verified that Next.js supplies `getServserSideProps` with the expected browser cookies which are available server side in `req.headers.cookie`. In other words if client has cookies, they end up into `req.headers.cookie` server side.

I wasn't implying that `document` object is available server side, but that cookie are available by default in `req.headers.cookie`.

Therefore I was suggesting "replicate" this behaviour in `next-page-tester` since JSDOM provides API for document.cookies.

Of course one could always hardcode cookies as `reqEnhancer`.

You mentioned that setting cookies like this before a test also feels weird:

```
document.cookie = 'TrackingId=123';
document.cookie = 'AnalyticsId=12345';
document.cookie = 'Foo=Bar';
```
In my opinion (just my opinion 😃 ) this is a bit more verbose, but it also feels natural since this would exactly replicate in tests the way a web application works. If I want to test that my protected page works as expected, I need to setup the test environment with necessary session cookies and so on. Exactly like would happen in browser or browser tests.

This would also cover the case of cookies being updated after the first rendering since we would re-read `document.cookies` at each render.

Do you see downsides to this approach?

**Comment by Meemaw** [comment]
@toomuchdesign we can go with Option 1 if you prefer that. It's true that it's closer to the DOM testing library philosophy. How did you confirm it works correctly now? I believe we need to make some changes.

e.g. If you go to the `cookies.test.tsx` and change first test (`'Should be able to access authenticated page by client side login with setting cookies'`) to use Option 1 it fails for me.

**Comment by toomuchdesign** [comment]
Hi @Meemaw,
if you agree I'll make an attempt on this PR. Luckily we have (lots of) tests, so we can get an impression if it works or not.

**Comment by Meemaw** [comment]
Sure, go ahead.

**Comment by toomuchdesign** [comment]
Hey @Meemaw,
here is a first draft of what we discussed earlier. I feel quite positive that it could be the right direction

### Open points 
There are a couple (minor?) points worth mentioning (which also apply to the previous implementation):

1. Shall we also append a `req.headers` object with the parsed cookie entries? (`ctx.req` seems to have it runtime, but it doesn't show up in typings, at least on my machine)
2. `getInitialProps` is supposed to have a `ctx` available only when run server side; currently we provide it with the same `ctx` object (and `ctx.req.headers.cookie`) no matter what

### Extra info
I tried to move all cookie-reated tests into the relevant `cookie` folder.

I moved `reqEnhancer` after `req.headers.cookie` setting, so that users can still optionally operate over appended data.

PS. I see we have randomly failing tests. Do you have any clue what could originate it?

🙌

**Comment by Meemaw** [comment]
> Hey @Meemaw,
> here is a first draft of what we discussed earlier. I feel quite positive that it could be the right direction
> 
> ### Open points
> There are a couple (minor?) points worth mentioning (which also apply to the previous implementation):
> 
> 1. Shall we also append a `req.headers` object with the parsed cookie entries? (`ctx.req` seems to have it runtime, but it doesn't show up in typings, at least on my machine)

Not sure if I understand. You mean append parsed cookie entries to `req.headers.cookie`? If I'm not mistaken that should  always be a string.

> 2. `getInitialProps` is supposed to have a `ctx` available only when run server side; currently we provide it with the same `ctx` object (and `ctx.req.headers.cookie`) no matter what

Yep, I would add that as a TODO and keep it for a separate PR. If I remember correctly `getInitialProps` executes first time on server and then always on client. So we also need to wrap it with `executeAsIfOnServer` on initial request.

> 3. `referer` doesn't currently get updated during navigation, right?

TBH, I haven't tested it properly. Let me look into it.

> 
> ### Extra info
> I tried to move all cookie-reated tests into the relevant `cookie` folder.
> 
> I moved `reqEnhancer` after `req.headers.cookie` setting, so that users can still optionally operate over appended data.
> 
> PS. I see we have randomly failing tests. Do you have any clue what could originate it?

I've seen them failing sometimes on CI (windows only) but have no idea what could be causing it. Were you able to reproduce it locally somehow?

> 
> 🙌

**Comment by toomuchdesign** [comment]
With regards to my 1st point about cookies, I wrote nonsense :). I meant that I found the parsed cookie object provided on runtime in `ctx.req.cookies`. I don't know how much standard it is (maybe Next.js uses [cookie-parser](https://www.npmjs.com/package/cookie-parser) under the hood) and whether we should consider implement it.

I agree about documentation, we should definitely dust it off and reshape it. I'd love to hear your take if you have any idea about this topic.

**Comment by toomuchdesign** [comment]
Made a quick refactoring to avoid wrapping `reqEnhancer` in `makePage` options overrides

**Comment by Meemaw** [comment]
Added some documentation here: https://github.com/toomuchdesign/next-page-tester/pull/47 so its easier to review.

**Comment by Meemaw** [comment]
> UPDATE: the random test failure seems to be due by `real-world-example` with error:

I think I have a suspect for this. Will look into it sometime this week after this is merged (its not related to this PR).

**Comment by Meemaw** [comment]
> 2. `getInitialProps` is supposed to have a `ctx` available only when run server side; currently we provide it with the same `ctx` object (and `ctx.req.headers.cookie`) no matter what

Fixed 🔧

**Comment by Meemaw** [comment]
Let's get this merged and released @toomuchdesign ? 🚀

**Review by toomuchdesign — APPROVED** [review]
Yep, let's release it!

I feel time has come to better organise tests. I like the approach we took recently of using separate pages folders instead of a global one, like I did initially.

I'd also like to consider a better division of responsibility among tests. Currently many tests are very general, which means that breaking a single feature usually leads to a massive test failure, making our maintainers lives harder :)

BTW @Meemaw, thank you so much for your effort here and for brining in ideas I'd never have came up with. It's been a big pleasure sharing all this with you. Unfortunately quite a rare pleasure in OS activity. :)

Catch up soon 🙌

**Pull request merged**
